Fix UniqueValidator with new `identifierFieldNames` parameter

soyuka opened the pull request:
Fix https://github.com/doctrine/DoctrineMongoDBBundle/issues/874

Breaking change in Symfony https://github.com/symfony/symfony/pull/38662

soyuka committed:
fixes https://github.com/doctrine/DoctrineMongoDBBundle/issues/874

GromNaN committed:
Add support for identifierFieldNames option

GromNaN committed:
Fix phpstan issues

GromNaN committed:
Use conditional trait to set the UniqueEntity::$service default value instead of overriding the contructor

GromNaN committed:
Redefining a property in a trait with a different default value is not allowed

GromNaN committed:
Remove useless phpcsSuppress

GromNaN commented:
I completely revised the implementation. In fact, we don't have to override the constructor to change only the default value of the `$service` parameter. We only need to redeclare the default value in the `$service` property.

We have to declare the class conditionally to keep the compatibility with Symfony 6.4 as the prop type has been [added in 7.0](https://github.com/symfony/symfony/commit/7ea24615d7b6f46a2af42cf860cd5f74f38eb7ad#diff-580bed00c9b884da104fb61d62ee451c36ef1ec4936c5a55b68aa4bd182c701b).

GromNaN committed:
PHPStan error is not ignorable

Property Doctrine\Bundle\MongoDBBundle\Validator\Constraints\Unique::$service overriding property
Symfony\Bridge\Doctrine\Validator\Constraints\UniqueEntity::$service (string) should also have native type
string.

https://github.com/phpstan/phpstan-src/blob/f4e4354045c2ed972ff3c4ad927fcad539a8a338/src/Rules/Properties/OverridingPropertyRule.php#L168-L176

jmikola commented on src/Validator/Constraints/Unique.php:
```diff
@@ -5,46 +5,36 @@
-#[Attribute(Attribute::TARGET_CLASS | Attribute::IS_REPEATABLE)]
-class Unique extends UniqueEntity
-{
+if ((new ReflectionProperty(UniqueEntity::class, 'service'))->hasType()) {
```
Noted that this is essentially a test for Symfony 7.0+ based on [symfony/symfony@7ea2461](https://github.com/symfony/symfony/commit/7ea24615d7b6f46a2af42cf860cd5f74f38eb7ad#diff-580bed00c9b884da104fb61d62ee451c36ef1ec4936c5a55b68aa4bd182c701b) adding parameter types to UniqueEntity.

jmikola commented on tests/Validator/Constraints/UniqueTest.php:
```diff
@@ -23,11 +25,71 @@ public function testWithDefaultProperty(): void
         self::assertInstanceOf(Unique::class, $constraint);
         self::assertSame(['email'], $constraint->fields);
         self::assertSame('doctrine_odm.mongodb.unique', $constraint->validatedBy());
+        self::assertNull($constraint->em);
```
I'm not questioning the correctness here, but why was a `$constraint->em` assertion introduced?

GromNaN replied:
I hit an issue while working on the previous implementation.

GromNaN replied:
It would have been so nice if this parameter was named `manager` to take into account the ODM.

GromNaN commented:
Thank you @soyuka for raising the issue again and proposing a solution.

soyuka commented:
thanks @GromNaN for taking this over my dirty patch :D
